src/swarm-engine/Swarm/Game/Scenario/Objective/WinCheck.hs

```diff
@@ -22,45 +23,38 @@ import Swarm.Game.Scenario.Objective.Logic as L
 
 -- | We have "won" if all of the "unwinnable" or remaining "incomplete" objectives are "optional".
 didWin :: ObjectiveCompletion -> Bool
-didWin oc = all _objectiveOptional $ incomplete buckets <> unwinnable buckets
- where
-  buckets = completionBuckets oc
+didWin oc =
+  and (oc ^.. incompleteObjectives . objectiveOptional)
+    && and (oc ^.. unwinnableObjectives . objectiveOptional)
```

Comment:
This new formulation is a bit less concise; in the original, we assemble a list of objectives and check that `all` of them are optional.  In the new code, we have to perform such a check twice independently.

Can we do something like ```incompleteObjectives `concatFold` unwinnableObjectives``` first?

Reply:
Oh, excellent point, thanks.  I think I tried something to simplify this at one point but it failed in a strange way (the game thought every scenario was instantly won no matter what).  But that was before I added `concatFold`.

The new version using `concatFold` is much shorter (and doesn't repeat code).